File: assets/js/hooks/mood_trajectory.js
import * as d3 from "d3"

/** PCA projection of daily mood dimensions to 2D — one scribble for the whole run. */
export const MoodTrajectory = {
  mounted() {
    this.data = { points: [], annotations: [], clusterColors: {}, clusterNames: {} }
    this.focusDate = null

    this.handleEvent("update-trajectory", (data) => {
      this.data = data
      this.render()
    })

    this.handleEvent("day-focus", ({ day }) => {
      this.focusDate = day?.date || null
      this.applyFocus()
    })

    this.handleEvent("update-trajectory-colors", ({ clusterByDate, clusterColors, clusterNames }) => {
      this.data.clusterColors = clusterColors
      this.data.clusterNames = clusterNames
      // Keep this.data.points current so resize re-renders with correct clusters
      this.data.points = this.data.points.map(p => ({ ...p, cluster: clusterByDate[p.date] ?? null }))

      const svg = d3.select(this.el).select("svg")
      if (svg.empty()) return

      // Use clusterByDate directly — bound datum still holds old cluster values
      svg.selectAll("line")
        .attr("stroke", d => {
          const cluster = clusterByDate[d[1].date] ?? null
          return (cluster && clusterColors[cluster]) || "#666"
        })

      // Mutate circle datum in-place so showTooltip sees the new cluster
      svg.selectAll("circle.dot")
        .each(function(d) { d.cluster = clusterByDate[d.date] ?? null })
        .attr("fill", d => (d.cluster && clusterColors[d.cluster]) || "#888")
    })

    this._onResize = () => this.render()
    window.addEventListener("resize", this._onResize)
  },

  destroyed() {
    window.removeEventListener("resize", this._onResize)
    if (this.tooltip) this.tooltip.remove()
  },

  render() {
    const { points, clusterColors } = this.data
    if (!points || points.length < 2) return

    this.el.innerHTML = ""
    this.el.style.position = "relative"

    const width = this.el.clientWidth || 900
    const height = 520
    const margin = { top: 24, right: 24, bottom: 24, left: 24 }
    const innerW = width - margin.left - margin.right
    const innerH = height - margin.top - margin.bottom

    this.dims = { innerW, innerH, margin }

    const { x, y } = this.computeScales()
    this.x = x
    this.y = y

    const svg = d3.select(this.el)
      .append("svg")
      .attr("width", width)
      .attr("height", height)
      .style("display", "block")

    const defs = svg.append("defs")

    const glow = defs.append("filter")
      .attr("id", "traj-glow")
      .attr("x", "-50%").attr("y", "-50%")
      .attr("width", "200%").attr("height", "200%")
    glow.append("feGaussianBlur").attr("stdDeviation", "1.6").attr("result", "blur")
    const merge = glow.append("feMerge")
    merge.append("feMergeNode").attr("in", "blur")
    merge.append("feMergeNode").attr("in", "SourceGraphic")

    const g = svg.append("g")
      .attr("transform", `translate(${margin.left}, ${margin.top})`)

    const segments = []
    for (let i = 1; i < points.length; i++) {
      segments.push([points[i - 1], points[i]])
    }

    this.lines = g.append("g")
      .selectAll("line")
      .data(segments)
      .join("line")
      .attr("x1", d => x(d[0].x))
      .attr("y1", d => y(d[0].y))
      .attr("x2", d => x(d[1].x))
      .attr("y2", d => y(d[1].y))
      .attr("stroke", d => (d[1].cluster && clusterColors[d[1].cluster]) || "#666")
      .attr("stroke-opacity", 0.55)
      .attr("stroke-width", 1.25)
      .attr("stroke-linecap", "round")

    this.tooltip = d3.select(this.el)
      .append("div")
      .attr("class", "trajectory-tooltip")
      .style("position", "absolute")
      .style("pointer-events", "none")
      .style("background", "rgba(10,10,26,0.92)")
      .style("color", "#eee")
      .style("padding", "8px 12px")
      .style("border-radius", "6px")
      .style("border", "1px solid rgba(255,255,255,0.08)")
      .style("font-size", "11px")
      .style("line-height", "1.5")
      .style("white-space", "nowrap")
      .style("z-index", "100")
      .style("opacity", 0)

    const hook = this

    this.circles = g.append("g")
      .selectAll("circle.dot")
      .data(points)
      .join("circle")
      .attr("class", "dot")
      .attr("cx", d => x(d.x))
      .attr("cy", d => y(d.y))
      .attr("r", 2.2)
      .attr("fill", d => (d.cluster && clusterColors[d.cluster]) || "#888")
      .attr("fill-opacity", 0.85)
      .attr("filter", "url(#traj-glow)")
      .attr("stroke-opacity", 0)

    this.hitCircles = g.append("g")
      .selectAll("circle.hit")
      .data(points)
      .join("circle")
      .attr("class", "hit")
      .attr("cx", d => x(d.x))
      .attr("cy", d => y(d.y))
      .attr("r", 7)
      .attr("fill", "transparent")
      .style("cursor", "crosshair")
      .on("mouseenter", function(event, d) { hook.showTooltip(event, d) })
      .on("mouseleave", function() { hook.hideTooltip() })
      .on("click", function(event, d) { hook.pushEvent("day_selected", { date: d.date }) })

    this.renderAnnotations(g, x, y)

    this.applyFocus()
  },

  /**
   * Personal milestones overlaid on the trajectory. Each annotation is
   * authored as `{date, label, note?}` in `Fugue.MoodLive.Annotations`. We
   * resolve each date to a projected (x, y), drop a small ring marker, and
   * place the label with a leader line. Layout is intentionally simple:
   * labels alternate above/below the point so successive annotations don't
   * collide horizontally. Tweak the file, refresh, see it move.
   *
   * Markers are interactive: hovering shows a tooltip with the label, date,
   * cluster, and optional note. Clicking selects the day (same as clicking
   * any trajectory point), so the calendar and day-detail panels react.
   */
  renderAnnotations(g, x, y) {
    const { points, annotations } = this.data
    if (!annotations || annotations.length === 0) return

    const byDate = new Map(points.map(p => [p.date, p]))

    const resolved = annotations
      .map((a, i) => {
        const point = byDate.get(a.date)
        if (!point) {
          console.warn(`[trajectory] annotation date not in dataset: ${a.date}`)
          return null
        }
        return { ...a, point, idx: i }
      })
      .filter(Boolean)
      .sort((a, b) => a.date < b.date ? -1 : 1)

    if (resolved.length === 0) return

    const hook = this
    const layer = g.append("g").attr("class", "trajectory-annotations")
    const labels = layer.append("g").attr("pointer-events", "none")
    const markers = layer.append("g")

    resolved.forEach((a, i) => {
      const px = x(a.point.x)
      const py = y(a.point.y)
      // Alternate above/below so adjacent labels stagger vertically.
      const above = i % 2 === 0
      const dy = above ? -28 : 28
      const dx = 14
      const lx = px + dx
      const ly = py + dy

      labels.append("line")
        .attr("x1", px).attr("y1", py)
        .attr("x2", lx).attr("y2", ly)
        .attr("stroke", "rgba(255,255,255,0.45)")
        .attr("stroke-width", 0.75)

      const text = labels.append("text")
        .attr("x", lx + 4)
        .attr("y", ly)
        .attr("text-anchor", "start")
        .attr("dominant-baseline", "middle")
        .attr("fill", "rgba(255,255,255,0.92)")
        .attr("font-size", "10.5px")
        .attr("font-family", "ui-serif, Georgia, serif")
        .attr("font-style", "italic")
        .text(a.label)

      // Dark halo behind the label for legibility against any cluster color.
      const bbox = text.node().getBBox()
      labels.insert("rect", "text:last-child")
        .attr("x", bbox.x - 3)
        .attr("y", bbox.y - 1)
        .attr("width", bbox.width + 6)
        .attr("height", bbox.height + 2)
        .attr("fill", "rgba(10,10,26,0.55)")
        .attr("rx", 2)

      markers.append("circle")
        .attr("cx", px).attr("cy", py)
        .attr("r", 4)
        .attr("fill", "none")
        .attr("stroke", "#fff")
        .attr("stroke-width", 1.25)
        .attr("stroke-opacity", 0.9)
        .attr("pointer-events", "none")

      // Generous transparent hit target. Sits above the day-level hit circles
      // so an annotated day's hover surfaces the milestone instead of the
      // generic cluster tooltip — that's almost always what the reader wants.
      markers.append("circle")
        .attr("cx", px).attr("cy", py)
        .attr("r", 10)
        .attr("fill", "transparent")
        .style("cursor", "pointer")
        .on("mouseenter", function(event) { hook.showAnnotationTooltip(event, a) })
        .on("mouseleave", function() { hook.hideTooltip() })
        .on("click", function() { hook.pushEvent("day_selected", { date: a.date }) })
    })
  },

  showAnnotationTooltip(event, a) {
    const cluster = a.point.cluster
    const clusterName = (cluster && this.data.clusterNames[cluster]) || null
    const color = (cluster && this.data.clusterColors[cluster]) || "#888"
    const noteHtml = a.note
      ? `<div style="margin-top:4px;color:#ddd;white-space:normal;max-width:240px">${a.note}</div>`
      : ""
    const clusterHtml = clusterName
      ? `<div style="margin-top:4px;color:${color}">${clusterName}</div>`
      : ""
    const html = `
      <div style="font-style:italic;font-family:ui-serif,Georgia,serif;font-size:12px">${a.label}</div>
      <div style="color:#888;font-size:10px;margin-top:1px">${a.date}</div>
      ${noteHtml}
      ${clusterHtml}
    `
    const rect = this.el.getBoundingClientRect()
    this.tooltip
      .html(html)
      .style("white-space", a.note ? "normal" : "nowrap")
      .style("top", (event.clientY - rect.top + 12) + "px")
      .style("left", Math.min(event.clientX - rect.left + 12, rect.width - 260) + "px")
      .style("opacity", 1)
  },

  computeScales() {
    const { points } = this.data
    const { innerW, innerH } = this.dims

    const xExt = d3.extent(points, p => p.x)
    const yExt = d3.extent(points, p => p.y)
    const xPad = (xExt[1] - xExt[0]) * 0.08 || 1
    const yPad = (yExt[1] - yExt[0]) * 0.08 || 1

    const dataW = (xExt[1] - xExt[0]) + 2 * xPad
    const dataH = (yExt[1] - yExt[0]) + 2 * yPad
    const fit = Math.min(innerW / dataW, innerH / dataH)
    const xMid = (xExt[0] + xExt[1]) / 2
    const yMid = (yExt[0] + yExt[1]) / 2

    const x = d3.scaleLinear()
      .domain([xMid - dataW / 2, xMid + dataW / 2])
      .range([(innerW - dataW * fit) / 2, (innerW + dataW * fit) / 2])

    const y = d3.scaleLinear()
      .domain([yMid - dataH / 2, yMid + dataH / 2])
      .range([(innerH + dataH * fit) / 2, (innerH - dataH * fit) / 2])

    return { x, y }
  },

  applyFocus() {
    if (!this.circles) return
    const target = this.focusDate

    if (!target) {
      this.circles.interrupt()
        .transition().duration(220)
        .attr("r", 2.2)
        .attr("stroke", "none")
        .attr("stroke-width", 0)
        .attr("stroke-opacity", 0)
      return
    }

    this.circles.filter(d => d.date !== target)
      .interrupt()
      .transition().duration(220)
      .attr("r", 2.2)
      .attr("stroke", "none")
      .attr("stroke-width", 0)
      .attr("stroke-opacity", 0)

    this.circles.filter(d => d.date === target)
      .interrupt()
      .attr("r", 9)
      .attr("stroke", "#fff")
      .attr("stroke-width", 2)
      .attr("stroke-opacity", 1)
      .transition().duration(900).ease(d3.easeCubicOut)
      .attr("r", 5)
      .attr("stroke-width", 1.25)
      .attr("stroke-opacity", 0.9)
  },

  showTooltip(event, d) {
    const name = (d.cluster && this.data.clusterNames[d.cluster]) || "—"
    const color = (d.cluster && this.data.clusterColors[d.cluster]) || "#888"
    const html = `<strong>${d.date}</strong><div style="color:${color}">${name}</div>`
    const rect = this.el.getBoundingClientRect()
    this.tooltip
      .html(html)
      .style("white-space", "nowrap")
      .style("top", (event.clientY - rect.top + 12) + "px")
      .style("left", Math.min(event.clientX - rect.left + 12, rect.width - 160) + "px")
      .style("opacity", 1)
  },

  hideTooltip() {
    if (this.tooltip) this.tooltip.style("opacity", 0)
  }
}

File: lib/fugue_web/live/mood_live/mood_trajectory.ex
defmodule FugueWeb.MoodLive.MoodTrajectory do
  @moduledoc """
  PCA-projected trajectory hero viz. Every day is a dot; consecutive days are
  connected with a cluster-colored segment. Replaces the former
  `MoodTrajectory` JS hook.

  Annotations (personal milestones) are overlaid with alternating above/below
  label positions and a dark stroke halo (`paint-order: stroke`) in place of
  the runtime-measured text background the JS used to compute.
  """

  use Phoenix.Component

  alias FugueWeb.MoodLive.{SvgMath, Tooltip}

  @width 900
  @height 520
  @m_top 24
  @m_right 24
  @m_bottom 24
  @m_left 24
  @inner_w @width - @m_left - @m_right
  @inner_h @height - @m_top - @m_bottom

  attr :points, :list, default: []
  attr :annotations, :list, default: []
  attr :cluster_colors, :map, default: %{}
  attr :cluster_names, :map, default: %{}
  attr :selected_day, :any, default: nil

  def trajectory(assigns) do
    focus_date =
      case assigns.selected_day do
        %{date: d} when is_binary(d) -> d
        _ -> nil
      end

    pts = assigns.points

    {segments, dots, annotation_markers} =
      if length(pts) < 2 do
        {[], [], []}
      else
        {x_fn, y_fn} = compute_scales(pts)

        segments =
          pts
          |> Enum.chunk_every(2, 1, :discard)
          |> Enum.map(fn [a, b] ->
            color = (b[:cluster] || b["cluster"]) |> color_for(assigns.cluster_colors)

            %{
              x1: x_fn.(val(a, :x)),
              y1: y_fn.(val(a, :y)),
              x2: x_fn.(val(b, :x)),
              y2: y_fn.(val(b, :y)),
              stroke: color
            }
          end)

        dots =
          Enum.map(pts, fn p ->
            cluster = val(p, :cluster)
            date = val(p, :date)
            cluster_name = (cluster && Map.get(assigns.cluster_names, cluster, cluster)) || "--"
            color = color_for(cluster, assigns.cluster_colors)

            %{
              cx: x_fn.(val(p, :x)),
              cy: y_fn.(val(p, :y)),
              date: date,
              color: color,
              focused?: date == focus_date,
              tooltip:
                ~s|<strong>#{date}</strong><div style="color:#{color}">#{cluster_name}</div>|
            }
          end)

        point_by_date = Map.new(pts, &{val(&1, :date), &1})

        annotation_markers =
          assigns.annotations
          |> Enum.map(&resolve_annotation(&1, point_by_date))
          |> Enum.reject(&is_nil/1)
          |> Enum.sort_by(& &1.date)
          |> Enum.with_index()
          |> Enum.map(fn {a, i} ->
            px = x_fn.(val(a.point, :x))
            py = y_fn.(val(a.point, :y))
            above? = rem(i, 2) == 0
            dy = if above?, do: -28, else: 28
            dx = 14

            cluster = val(a.point, :cluster)
            color = color_for(cluster, assigns.cluster_colors)
            cluster_name = cluster && Map.get(assigns.cluster_names, cluster, cluster)

            %{
              px: px,
              py: py,
              label_x: px + dx,
              label_y: py + dy,
              text_x: px + dx + 4,
              text_y: py + dy,
              label: a.label,
              date: a.date,
              tooltip: annotation_tooltip(a, cluster_name, color)
            }
          end)

        {segments, dots, annotation_markers}
      end

    assigns =
      assign(assigns,
        segments: segments,
        dots: dots,
        annotation_markers: annotation_markers,
        svg_width: @width,
        svg_height: @height,
        g_transform: "translate(#{@m_left},#{@m_top})"
      )

    ~H"""
    <Tooltip.container id="mood-trajectory" style="width: 100%;">
      <svg
        viewBox={"0 0 #{@svg_width} #{@svg_height}"}
        preserveAspectRatio="xMidYMid meet"
        style="display: block; width: 100%;"
      >
        <defs>
          <filter id="traj-glow" x="-50%" y="-50%" width="200%" height="200%">
            <feGaussianBlur stdDeviation="1.6" result="blur" />
            <feMerge>
              <feMergeNode in="blur" />
              <feMergeNode in="SourceGraphic" />
            </feMerge>
          </filter>
        </defs>

        <g transform={@g_transform}>
          <g>
            <%= for seg <- @segments do %>
              <line
                x1={SvgMath.fmt(seg.x1)}
                y1={SvgMath.fmt(seg.y1)}
                x2={SvgMath.fmt(seg.x2)}
                y2={SvgMath.fmt(seg.y2)}
                stroke={seg.stroke}
                stroke-opacity="0.55"
                stroke-width="1.25"
                stroke-linecap="round"
              />
            <% end %>
          </g>

          <g>
            <%= for d <- @dots do %>
              <circle
                cx={SvgMath.fmt(d.cx)}
                cy={SvgMath.fmt(d.cy)}
                r={if d.focused?, do: "5", else: "2.2"}
                fill={d.color}
                fill-opacity="0.85"
                filter="url(#traj-glow)"
                stroke={if d.focused?, do: "#fff", else: "none"}
                stroke-width={if d.focused?, do: "1.25", else: "0"}
                stroke-opacity={if d.focused?, do: "0.9", else: "0"}
              />
            <% end %>
          </g>

          <g>
            <%= for d <- @dots do %>
              <circle
                class="traj-hit"
                cx={SvgMath.fmt(d.cx)}
                cy={SvgMath.fmt(d.cy)}
                r="7"
                fill="transparent"
                data-tooltip={d.tooltip}
                phx-click="day_selected"
                phx-value-date={d.date}
                style="cursor: crosshair;"
              />
            <% end %>
          </g>

          <g class="trajectory-annotations">
            <%= for a <- @annotation_markers do %>
              <line
                x1={SvgMath.fmt(a.px)}
                y1={SvgMath.fmt(a.py)}
                x2={SvgMath.fmt(a.label_x)}
                y2={SvgMath.fmt(a.label_y)}
                stroke="rgba(255,255,255,0.45)"
                stroke-width="0.75"
                pointer-events="none"
              />
              <text
                x={SvgMath.fmt(a.text_x)}
                y={SvgMath.fmt(a.text_y)}
                text-anchor="start"
                dominant-baseline="middle"
                font-size="10.5px"
                font-family="ui-serif, Georgia, serif"
                font-style="italic"
                paint-order="stroke"
                stroke="rgba(10,10,26,0.65)"
                stroke-width="3"
                stroke-linejoin="round"
                fill="rgba(255,255,255,0.92)"
                pointer-events="none"
              >
                {a.label}
              </text>
              <circle
                cx={SvgMath.fmt(a.px)}
                cy={SvgMath.fmt(a.py)}
                r="4"
                fill="none"
                stroke="#fff"
                stroke-width="1.25"
                stroke-opacity="0.9"
                pointer-events="none"
              />
              <circle
                class="annotation-hit"
                cx={SvgMath.fmt(a.px)}
                cy={SvgMath.fmt(a.py)}
                r="10"
                fill="transparent"
                data-tooltip={a.tooltip}
                phx-click="day_selected"
                phx-value-date={a.date}
                style="cursor: pointer;"
              />
            <% end %>
          </g>
        </g>
      </svg>
    </Tooltip.container>
    """
  end

  # --- Geometry helpers ---

  defp compute_scales(points) do
    xs = Enum.map(points, &val(&1, :x)) |> Enum.map(&(&1 * 1.0))
    ys = Enum.map(points, &val(&1, :y)) |> Enum.map(&(&1 * 1.0))

    x_min = Enum.min(xs)
    x_max = Enum.max(xs)
    y_min = Enum.min(ys)
    y_max = Enum.max(ys)

    x_pad = max((x_max - x_min) * 0.08, 1.0)
    y_pad = max((y_max - y_min) * 0.08, 1.0)

    data_w = x_max - x_min + 2 * x_pad
    data_h = y_max - y_min + 2 * y_pad
    fit = min(@inner_w / data_w, @inner_h / data_h)

    x_mid = (x_min + x_max) / 2
    y_mid = (y_min + y_max) / 2

    x_domain_lo = x_mid - data_w / 2
    x_range_lo = (@inner_w - data_w * fit) / 2

    y_domain_hi = y_mid + data_h / 2
    y_range_lo = (@inner_h - data_h * fit) / 2

    x_fn = fn v -> x_range_lo + (v - x_domain_lo) * fit end
    # SVG y grows downward; trajectory's y_max should map to the TOP of the inner area.
    y_fn = fn v -> y_range_lo + (y_domain_hi - v) * fit end

    {x_fn, y_fn}
  end

  defp color_for(nil, _), do: "#888"
  defp color_for(cluster, colors), do: Map.get(colors, cluster, "#888")

  defp val(p, key) when is_map(p) do
    Map.get(p, key) || Map.get(p, to_string(key))
  end

  defp resolve_annotation(a, point_by_date) do
    date = val(a, :date)

    case Map.get(point_by_date, date) do
      nil ->
        nil

      point ->
        %{
          date: date,
          label: val(a, :label),
          note: val(a, :note),
          point: point
        }
    end
  end

  defp annotation_tooltip(a, cluster_name, color) do
    note_html =
      case a.note do
        note when is_binary(note) and note != "" ->
          ~s|<div style="margin-top:4px;color:#ddd;white-space:normal;max-width:240px">#{note}</div>|

        _ ->
          ""
      end

    cluster_html =
      if cluster_name do
        ~s|<div style="margin-top:4px;color:#{color}">#{cluster_name}</div>|
      else
        ""
      end

    ~s|<div style="font-style:italic;font-family:ui-serif,Georgia,serif;font-size:12px">#{a.label}</div>| <>
      ~s|<div style="color:#888;font-size:10px;margin-top:1px">#{a.date}</div>| <>
      note_html <> cluster_html
  end
end
